src/main/java/org/jenkinsci/plugins/workflow/multibranch/DurabilityHintBranchProperty.java

```diff
@@ -0,0 +1,82 @@
+package org.jenkinsci.plugins.workflow.multibranch;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.Extension;
+import hudson.model.Job;
+import hudson.model.JobProperty;
+import hudson.model.Run;
+import jenkins.branch.BranchProperty;
+import jenkins.branch.BranchPropertyDescriptor;
+import jenkins.branch.JobDecorator;
+import org.jenkinsci.plugins.workflow.flow.FlowDurabilityHint;
+import org.jenkinsci.plugins.workflow.flow.GlobalDefaultFlowDurabilityLevel;
+import org.jenkinsci.plugins.workflow.job.WorkflowJob;
+import org.jenkinsci.plugins.workflow.job.properties.DurabilityHintJobProperty;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+import javax.annotation.Nonnull;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * Branch property so we can define per-branch durability policies, i.e. so feature branches aren't built durably but master is.
+ * NEEDS WORKFLOW-JOB-PROPERTY
+ * @author Sam Van Oort
+ */
+@Restricted(NoExternalUse.class)
+public class DurabilityHintBranchProperty extends BranchProperty {
+
+    private final FlowDurabilityHint hint;
+
+    public FlowDurabilityHint getHint() {
+        return hint;
+    }
+
+    @DataBoundConstructor
+    public DurabilityHintBranchProperty(@Nonnull FlowDurabilityHint hint) {
+        this.hint = hint;
+    }
+
+    @Override
+    public final <P extends Job<P, B>, B extends Run<P, B>> JobDecorator<P, B> jobDecorator(Class<P> clazz) {
+        if (!clazz.isAssignableFrom(WorkflowJob.class)) {
+            return null;
+        }
+        return new JobDecorator<P, B>() {
+            @NonNull
+            @Override
+            public List<JobProperty<? super P>> jobProperties(
+                    @NonNull List<JobProperty<? super P>> jobProperties) {
+                List<JobProperty<? super P>> result = asArrayList(jobProperties);
+                for (Iterator<JobProperty<? super P>> iterator = result.iterator(); iterator.hasNext(); ) {
+                    JobProperty<? super P> p = iterator.next();
+                    if (p instanceof DurabilityHintJobProperty) {
+                        iterator.remove();
+                    }
+                }
+                if (getHint() != null) {
+                    result.add((JobProperty)(new DurabilityHintJobProperty(getHint())));
```

Comment:
This speaks to the need for something like

```java
public interface FlowDurabilityHintProvider extends ExtensionPoint
    @CheckForNull FlowDurabilityHint suggestFor(@Nonnull FlowExecution x);
    static @Nonnull FlowDurabilityHint suggestedFor(@Nonnull FlowExecution x) {
        for (FlowDurabilityHintProvider p : ExtensionList.lookup(FlowDurabilityHintProvider.class)) {
            FlowDurabilityHint h = p.suggestFor(x);
            if (h != null) {
                return h;
            }
            return FlowDurabilityHint.MAX_SURVIVABILITY;
        }
    }
}
```

with implementations

* job property
* branch property
* (optionally) folder property
* global

Reply:
A loose adaption of this is now applied (simplified somewhat).